### ISSUE TITLE
System.Net.Sockets.SocketException (0x80004005): An existing connection was forc ibly closed by the remote host

### ISSUE THREAD

**Issue by peske** [zeromq/netmq#565]
As far as I understand ZeroMQ should hide all underlying socket communication, thus should be resilient to any errors that may happen there. But obviously it is not the case, at least when NetMQ is used. I've created simple REQ/REP pattern for testing purposes (by using `RequestSocket` and `ResponseSocket` instances), and discovered that calling `RequestSocket.Dispose()` method causes the following exception at `ResponseSocket` side:

```
System.Net.Sockets.SocketException (0x80004005): An existing connection was forc
ibly closed by the remote host
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size,
 SocketFlags socketFlags)
   at NetMQ.Core.Mailbox.TryRecv(Int32 timeout, Command& command)
   at NetMQ.Core.SocketBase.ProcessCommands(Int32 timeout, Boolean throttle)
   at NetMQ.Core.SocketBase.TryRecv(Msg& msg, TimeSpan timeout)
   at NetMQ.NetMQSocket.TryReceive(Msg& msg, TimeSpan timeout)
   at NetMQ.ReceivingSocketExtensions.Receive(IReceivingSocket socket, Msg& msg)

   at NetMQ.ReceivingSocketExtensions.ReceiveMultipartMessage(IReceivingSocket s
ocket, Int32 expectedFrameCount)
   ...
```

Is this a bug or expected behavior? Am I doing something wrong?

Thanks!

**Comment by tatianagrabcad** [comment]
I am getting same issue on Windows 7: CurrentDomain.UnhandledException 
System.Net.Sockets.SocketException (0x80004005): An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at NetMQ.zmq.Mailbox.Recv(Int32 timeout)
   at NetMQ.zmq.Reaper.InEvent()
   at NetMQ.zmq.Poller.Loop()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

**Comment by peske** [comment]
I've forgot to say one thing that can be important: `RequestSocket` instance is disposed from another thread (not the same one used for connecting / sending / receiving). But in any case I'm pretty sure that `ResponseSocket` (which usually plays _server_ role) should not fail because of any error that happened at `RequestSocket` (which usually plays _client_ role). It would be very wrong if any error at any of _clients_ can crash the _server_.

**Comment by joaoantunes** [comment]
RequestSocket instance is disposed from another thread

**Comment by peske** [comment]
I've located the problem, so first I need to apologize! I've misinterpreted the problem in my original post - the exception does not happen when _client_ disposes, but when _server_ (`ResponseSocket`) is disposed from another thread while still being in blocking read in the original thread. 

I've misinterpreted things because disposals of _client_ and _server_ are next to one another (the same `using` block), and error message confused me because of part "...was forcibly closed by the **remote** host..." Since the error happens on the _server_ I've assumed that it is caused by disposing _client_ (remote).

Anyway it is my mistake. Exception thrown does make sense (and even should happen) in these circumstances. 

Sorry and thanks!
